content/blog/2019-09-25-floats-static-arrays.md

```diff
@@ -0,0 +1,67 @@
++++
+title = "Floating Points and Fixed-Length Arrays in Bril"
+extra.author = "Dietrich Geisler"
+extra.bio = """
+  [Dietrich Geisler](https://www.cs.cornell.edu/~dgeisler/) is a 3rd year PhD student researching Language Design and Compilers.  Enjoys gaming and climbing.
+"""
++++
+
+### Introduction
+
+My goal for this project was to add floating points and fixed-length arrays to [Bril](https://github.com/sampsyo/bril).  The intention behind this decision is to promote Bril as a lower-level intermediate language, with fewer type abstractions than Javascript without the complexity of representation of LLVM.  The hope is that, by writing the Braid IR to provide a _minimal_ but _descriptive_ set of numeric operations, users will be able to explore optimiziations without sacrificing visibility of low-level computational operations.
+
+These goals were mostly successful at the level of Bril semantics, parsing, and interpretation.  What was more difficult than expected (and less successful) were my attempts to model TypeScript values and arrays with [TS2Bril](https://github.com/sampsyo/bril/blob/master/bril-ts/ts2bril.ts).  Before exploring the limitations of the Bril representation, however, we must examine the floating point and array semantics of Bril more formally.
+
+### Floating Points in Bril
+
+Bril now supports the types `double` and `float`.  These represent, respectively,  the stadard IEEE-754 [double-precision](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) and [single-precision](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) floating-point formats.  These types come equipped with the basic arithmetic operations `fadd`, `fsub`, `fmult`, and `fdiv`, along with the comparitives `feq`, `flt`, `fgt`, `fle`, and `fge`.  The introduction of these operations is intended to highlight that floating-point and integer operations should be treated as fundamentally different objects.
+
+In Bril, floating-point precisions should not be mixed.  It is expected, for instance, that the following code will produce a static exception:
+```
+v0: double = const 5;
+v1: float = const 5;
+v2: float = fadd v0 v1
+```
+Similarly, mixing floating-points and integers in Bril is undefined, as well as using floating-point arithmetic operations on integers or vice-versa.  In a future project, I intend to explore the consequences of relaxing these semantic requirements, particularly when mixing floating-point precision.
+
+### Bril Fixed-Length Arrays
+
+Bril now supports types of the form `array=type[size]`, where `type = | int | bool | float | double | array`.  Note the inclusion of type recursion in this definition, which permits multi-dimensional arrays.  Arrays in Bril must contain elements all of the same type, and the size of arrays cannot be changed after initialization.
```

Comment:
Seems like the type is just `type[size]`, right? What's up with the `array=` prefix?

Reply:
Looks like there's an extra leading `|` in your list for `type`. Also, I might just say that `type` is any type, including an array, instead of listing all the currently-existing types.